### PR TITLE
Add deep linking support

### PR DESCRIPTION
**Pull request by csduarte**
#### Summary
Allow permalinks to open the mobile app.

The app checks for platform specific configuration on app install. If no configuration is found then the deep linking code sits silently and permalinks act as just regular links.

##### Setup for iOS
- Create an `apple-app-site-association` file in the `.well-known` directory at the root of your server. It should be accessible by navigating to `https://<your-site-name>/.well-known/apple-app-site-association`. *** There should NOT be a file extension ***
- In order to handle permalink deep links paste the following JSON into the `apple-app-site-association` file. Make sure to place your app ID in the `appID` property:
```
{
    "applinks": {
        "apps": [],
        "details": [
            {
                "appID": "<your-app-id-here>",
                "paths": ["**/pl/*"]
            }
        ]
    }
}
```
- Make sure to add the associated domains entitlement to your app via the Apple developer portal
- Add an entitlement that specifies the domains your app supports via the Xcode entitlements manager.

Before installing the app with the new entitlement make sure that you can view the contents of the `apple-app-site-association` file via a browser by navigating to `https://<your-site-name>/.well-known/apple-app-site-association`. The app will check for this file on install and if found will allow outside permalinks to open the app.

Official documentation can be found using the following link:
- [iOS universal links](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppSearch/UniversalLinks.html)

##### Setup for Android
Configuring deep links for Android is done much easier when using the App Links Assistant in Android Studio. Please refer to the following link for adding deep links via the App Links Assistant:
- [Android App Links Assistant](https://developer.android.com/studio/write/app-link-indexing)

#### Screenshots
kapture 2018-06-22 at 14 41 11

### CODE REVIEW

**Comment by enahum** [comment]
@jarredwitt is there documentation needed for this?

**Comment by jarredwitt** [comment]
@enahum updated the PR description to include some instructions for setup. The setup is pretty specific for each platform since it deals with the app ID's as well as some required file uploads to the server. The good thing about the setup is that it sits silently in the background if a configuration is not found. Let me know if you need help/have issues testing.

**Comment by enahum** [comment]
It would be great to submit the PR to the docs and the developer docs ;)

**Comment by enahum** [comment]
Will this work with multiple servers if the app id is the one we provide or people wanting to use this will need to build their own apps?

**Comment by jarredwitt** [comment]
@enahum sorry for the late response, not sure on the multiple servers front. My guess is maybe? For iOS the `apple-app-site-association` file it only requires that the `appID` be present so I think if users are using the build from the app store then they would only need to upload the `apple-app-site-association` file to their server and I think it'll work. Not sure on Android, but I think if the file is signed using the MM keystore then it should work the same way.

**Comment by jarredwitt** [comment]
@enahum actually forgot that comment. It'll require users to build their own app if they want to support other servers since you have to specify the site url in the entitlements file at build time for iOS and Android.

File: app/screens/channel/channel_post_list/channel_post_list.js [hunk 1]
```diff
@@ -58,6 +58,15 @@ export default class ChannelPostList extends PureComponent {
         this.contentHeight = 0;
     }
 
+    componentDidMount() {
+        this.mounted = true;
```
Comment (enahum) [inline app/screens/channel/channel_post_list/channel_post_list.js]: is there a reason for this?

Comment (jarredwitt) [reply app/screens/channel/channel_post_list/channel_post_list.js]: Yes, I was getting some setState on unmounted component warnings from this:
https://github.com/uber-uchat/mattermost-mobile/blob/ca1659c65835eb0c3a0d632811eaa860c1585f73/app/screens/channel/channel_post_list/channel_post_list.js#L63

**Pull request merged**